core/src/main/java/hudson/model/Job.java

```diff
@@ -424,15 +427,24 @@ public synchronized void updateNextBuildNumber(int next) throws IOException {
     }
 
     /**
-     * Returns the configured build discarder for this job, or null if none.
+     * Returns the configured build discarder for this job, via {@link BuildDiscarderProperty}, or null if none.
      */
     public BuildDiscarder getBuildDiscarder() {
-        return logRotator;
+        BuildDiscarderProperty prop = getProperty(BuildDiscarderProperty.class);
```

Comment:
I suppose. Probably irrelevant in practice since the old getter and setter methods are now only called from migration code, so typically never. Have considered deprecating them and `logRotate` and making it call itself as a `RunListener.onFinalized`, so that eventually the whole `BuildDiscarder` system could be independent (no more hooks needed in `Job`/`Run`).